src/ctia/lib/riemann.clj

```diff
@@ -0,0 +1,124 @@
+         :mask (str (= "Mask" (get-in request [:headers "x-client-app"])))
+         :identity (:identity request)
+         :jwt (:jwt request)}
+        extra-fields))
+
+(defn ms-elapsed
+  "provide how much ms were elapsed since `nano-start`."
+  [nano-start]
+  (/ (- (System/nanoTime) nano-start) 1000000.0))
+
+(defn- send-request-metrics [send-event-fn request extra-fields]
+  (let [event (request->event request extra-fields)]
+    (try
+      (send-event-fn event)
+      (catch Exception e
+        (log/warnf "A Problem occured while sending request metrics event:\n\n%s\n\nException:\n%s"
+                   (utils/safe-pprint event)
+                   (utils/safe-pprint e))
+        (when (nil? send-event-fn)
+          (log/warn "The send-event-fn looks empty. This is certainly due to a configuration problem or perhaps simply a code bug."))))))
+
+(defn wrap-request-metrics [handler service-name send-event-fn]
+  (fn [request]
+    (let [start (System/nanoTime)]
+      (try
+        (when-let [response (handler request)]
+          (let [ms (ms-elapsed start)]
+            (send-request-metrics send-event-fn request
+                                  {:metric ms
+                                   :tags ["ctia" "http"]
+                                   :description (str "Response took "
+                                                     (.format (java.text.DecimalFormat. "#.##")
+                                                              (/ ms 1000))
+                                                     " seconds")
+                                   :service service-name
+                                   :status (str (:status response))}))
+          response)
+        (catch ExceptionInfo e
+          (let [data (ex-data e)
+                ex-type (:type data)
+                evt {:metric (ms-elapsed start)
+                     :service (str service-name " error")
+                     :description (.getMessage e)
+                     :error "true"
+                     :status (condp = ex-type
+                               :ring.util.http-response/response (get-in data [:response :status])
+                               :compojure.api.exception/response-validation "500"
+                               :compojure.api.exception/request-validation "400"
+                               "500")}]
+            (send-request-metrics send-event-fn request evt))
+          (throw e))
+        (catch Throwable e
+          (send-request-metrics send-event-fn request
+                                {:metric (ms-elapsed start)
+                                 :service (str service-name " error")
+                                 :description (.getMessage e)
+                                 :stacktrace (.getStackTrace e)
+                                 :status "500"
+                                 :error "true"})
+          (throw e))))))
+
+(defn prepare-event [event]
+  {:user-id (get-in event [:user :id])})
```

Comment:
The `prepare-event` in IROH https://github.com/threatgrid/iroh/blob/master/services/riemann-reporter/src/riemann_reporter/core.clj#L172 does not only take care of adding useful metas it takes care of sending flattened hash-map, remove potential sensible informations that should never be visible in the logs (like API keys, raw JWT, etc...).

You should probably copy it. Except that `find-and-add-metas` function that you could get rid of by taking care of declaring manually `user-id`, `user-name`, `user-email`, `org-id`, `org-name`, `idp-id`, `client-id`, `client-name` and `trace-id` directly by using the `identity` value of the request.

It will be very important to have the same structure between IROH and CTIA logs otherwise it will be a nightmare for the people looking at the logs.

Reply:
Ok, I copied everything pretty much verbatim. I don't follow about the `identity` value, could you help me change `extract-infos-from-event` to follow your suggestion?

Reply:
So `extract-infos-from-event` is a complex function that try to retrieve `user-id` that could be hidden in deeper keys of a nested hash-map. For example, we had both:

```clojure
(send-event {:jwt {"https://schemas.cisco.com/iroh/identity/claims/user/id" "foo"
             "https://schemas.cisco.com/iroh/identity/claims/user/name" "John Doe"
             "https://schemas.cisco.com/iroh/identity/claims/org/id" "..."
             :some-value {:inner "foo" ,,,}})

;; and

(send-event {:identity {:user {:id "foo" :name "John Doe",,,}
                               :org {:id "..." ,,,}})

;;

(send-event {:user-id "foo" :org-id "..." ,,,})
```

So if you would like to search all event for some `user-id` the kibana query became complex. So to simplify that the `extract-infos-from-event` try to search for all value that could be `user-id`. This is not perfect, some events might have multiple potentially different value for `user-id`. But it works most the time. And the situation was bad because we haven't imposed any event format convention.

As you are starting new in CTIA, I think you can enforce such a convention.
So, right now, we have try to generate all the fields I mentioned (`user-id`, `user-name`, etc...).
You can get those by looking at the `:identity` of the ring request.

So the `ring-jwt-middleware/wrap-jwt-auth-fn` middleware take a ring request (a map) and add an `:identity` key containing an `Identity`:

https://github.com/threatgrid/iroh/blob/803120d750f3be604470a0965205128ea27f7f1a/lib/iroh-web/src/iroh_web/request_context.clj#L18--L31

From this you should be able to simply write something like

```clojure
(defn identity->event-fields [req-identity]
    {:user-id (get-in req-identity [:user :id]
     ,,,})
```

And use this function when generating the `event` from the ring request.

Reply:
Another note, if you keep using the `extract-infos-from-event` it will work but might just not be optimal.
And if you start to send big events it might potentially affect the response time.
In IROH there is a test trying the check that it this function is fast enough. 

So if you don't feel comfortable to change it right away you could keep it and improve the code later.